### ISSUE TITLE
[miio >= 0.15.1] Mi Vacuum Cleaner Robot connection issue

### ISSUE THREAD

**Issue by jghaanstra** [aholstenson/miio#94]
This is a new issue for the Mi Robot Vacuum Cleaner. It seems for some users it works and for others it doesnt (anymore). I had it working fine until I installed firmware 3.3.9_003077 on the robot, it then stopped working. I was hoping the new miio version would fix the issue but for me it doesnt. So this Github issue is intended to gather as much information as possible that can explain the difference in results. It would be nice if more people could chip in to see if we can find a pattern.

- **Robot firmware:** 3.3.9_003094
- **Xiaomi server:** China Mainland
- **Router:** Asus AC68U
- **DNS:** Google (8.8.4.4) and OpenDNS (208.67.220.220)
- **Has it worked previously:** Yes
- **Token retrieved:** Rooted Android phone and viewed the table devicerecord in miio2.db to get token after connecting to WiFi
- **Other**: I also see this in the devicerecord table in the miio2.db field extraInfo: {"isSetPincode":0,"fw_version":"3.3.9_003094","needVerifyCode":0,"isPasswordEncrypt":0,"is_factory":false}

I'm including DNS info as some have suggested that timeout issues are related to certain DNS services being blocked by Xiaomi. It does not seem relevant in this case as it's LAN traffic but who knows ...

When using miio.device() in the example.js with the token added manually in the code the following error is thrown:
```
Error occurred: { Error: Could not connect to device, token might be wrong
    at enrichPromise.promise.then.then.catch.err (E:\Localhost\bin\nodejs\node_modules\miio\lib\network.js:310:16)
    at <anonymous>
  code: 'connection-failure',
  device:
   DeviceInfo {
     parent:
      Network {
        domain: null,
        _events: {},
        _eventsCount: 0,
        _maxListeners: undefined,
        packet: [Object],
        addresses: [Object],
        devices: [Object],
        references: 0,
        debug: [Object],
        _socket: null },
     packet:
      Packet {
        discovery: false,
        header: <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 67 ae 2e 2a f9 75 83 58 5f c0 3e 3c 34 a2 7b ea d1 eb a7>,
        _serverStampTime: 1516744227679,
        _token: <Buffer 41 64 56 45 39 73 77 61 62 30 65 77 64 43 58 58>,
        _tokenKey: <Buffer 3a 25 f0 ae fb 1c 53 0f a5 25 a6 7e 85 53 fa 1d>,
        _tokenIV: <Buffer 7a 05 4a 8a 8a 00 74 bb 2c 7c 1c 0c 2f 82 c0 45>,
        data: <Buffer 7b 22 69 64 22 3a 31 2c 22 6d 65 74 68 6f 64 22 3a 22 6d 69 49 4f 2e 69 6e 66 6f 22 2c 22 70 61 72 61 6d 73 22 3a 5b 5d 7d>,
        _serverStamp: 1516744228 },
     address: '192.168.0.133',
     port: 54321,
     promises: Map {},
     lastId: 1,
     id: 51850703,
     debug:
      { [Function: debug]
        namespace: 'thing:miio:51850703',
        enabled: false,
        useColors: true,
        color: 203,
        destroy: [Function: destroy],
        inspectOpts: {} },
     tokenChanged: true,
     handshakeResolve: null,
     handshakeTimeout: null,
     handshakePromise: null,
     autoToken: true,
     enrichPromise: null,
     enriched: true } }
```

**Comment by aholstenson** [comment]
Thank you for the details! As you said it would be great if more people could add details about their own setups and we'll see if we can work out what's wrong together.

**Comment by jghaanstra** [comment]
Here is some extra info. I had to remove tokens.json from my Windows profile to have miio CLI-tool discover the robot again. Using the tokens update command does not seem to regenerate a new tokens.json anymore. But when setting the token manually in the CLI-tool or in example.js I still recieve the error `Could not connect to device, token might be wrong` . I have reset my vacuum cleaner, reconnected it to my WiFi, retrieved the new token again from miio2.db but to no avail. The error remains. Below is the complete debug log. I'm still clueless as to why this is not working for my vacuum cleaner.

```
  miio:network Grabbing reference to network +0ms
  miio:network Making network active, creating socket +4ms
  miio:packet -> <Buffer 21 31 00 20 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 68 6f 9b ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +80ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 68 6f 9b ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +3ms
  thing:miio:pending <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  thing:miio:51850703 Identifier of device updated +0ms
  thing:miio:51850703 -> (5) {"id":1,"method":"miIO.info","params":[]} +1ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f 9b 79 ba 03 6a 29 c2 74 74 52 08 2c 26 23 fd ac 09> +5ms
  thing:miio:51850703 -> (4) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f 9d 9c ee 0b 87 ed 4c 27 20 1d c0 a6 d8 82 a9 7a 3e> +2s
  thing:miio:51850703 -> (3) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f 9f 4c 7c 1b c2 b2 8c 8f 7c bb a7 d9 6e 5c 15 b0 47> +2s
  thing:miio:51850703 -> (2) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f a1 69 2a 69 eb 48 d6 40 db c8 9f 00 ec 23 6d 83 2c> +2s
  thing:miio:51850703 -> (1) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f a3 9f 86 a2 6b 66 52 14 e1 67 9d 5e f6 58 b4 bc 00> +2s
  thing:miio:51850703 -> (0) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f a5 ae 5e 1e 12 1d cc 9b 9d 92 ae a4 75 51 e3 c1 fb> +2s
  miio:network Releasing reference to network +12s
  miio:network Network no longer active, destroying socket +1ms
Error occurred: { Error: Could not connect to device, token might be wrong
    at enrichPromise.promise.then.then.catch.err (E:\Localhost\bin\nodejs\node_modules\miio\lib\network.js:310:16)
    at <anonymous>
  code: 'connection-failure',
  device:
   DeviceInfo {
     parent:
      Network {
        domain: null,
        _events: {},
        _eventsCount: 0,
        _maxListeners: undefined,
        packet: [Object],
        addresses: [Object],
        devices: [Object],
        references: 0,
        debug: [Object],
        _socket: null },
     packet:
      Packet {
        discovery: false,
        header: <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 68 6f a5 ae 5e 1e 12 1d cc 9b 9d 92 ae a4 75 51 e3 c1 fb>,
        _serverStampTime: 1516793754661,
        _token: <Buffer 4d 70 70 56 32 37 54 56 45 42 56 6f 34 39 62 74>,
        _tokenKey: <Buffer eb 02 18 17 c6 9e a5 a3 37 30 be d4 95 27 d8 6e>,
        _tokenIV: <Buffer 00 ba 91 29 24 1a 5f 62 25 c4 7c 1a 93 ee 53 c8>,
        data: <Buffer 7b 22 69 64 22 3a 31 2c 22 6d 65 74 68 6f 64 22 3a 22 6d 69 49 4f 2e 69 6e 66 6f 22 2c 22 70 61 72 61 6d 73 22 3a 5b 5d 7d>,
        _serverStamp: 1516793755 },
     address: '192.168.0.133',
     port: 54321,
     promises: Map {},
     lastId: 1,
     id: 51850703,
     debug:
      { [Function: debug]
        namespace: 'thing:miio:51850703',
        enabled: true,
        useColors: true,
        color: 203,
        destroy: [Function: destroy],
        inspectOpts: {},
        diff: 2004,
        prev: 1516793762678,
        curr: 1516793764682 },
     tokenChanged: true,
     handshakeResolve: null,
     handshakeTimeout: null,
     handshakePromise: null,
     autoToken: true,
     enrichPromise: null,
     enriched: true } }
```

**Comment by aholstenson** [comment]
Thanks! That gives me a few ideas. It looks like the library connects to the vacuum sometimes then. Which is confusing. The dwaal-error is unrelated, it was for an experimental naming feature, but it doesn't play nice with Windows so I have removed it in 0.15.2.

But if it works sometimes it might an issue with the vacuum sleeping or otherwise not responding to network requests. But I would think that it wouldn't respond to the initial handshake if that was case.

I'm going to look into two things based on your log:
1) Check that the token is correctly set before the first call to `miIO.info`
2) The delay and number of retries for the initial call to `miIO.info` should be increased to see if it helps

**Comment by 350d** [comment]
I've installed latest `miio` on my new Pi setup and my Mi Robot stops working too. Here is my error report https://github.com/seikan/homebridge-xiaomi-mi-robot-vacuum/issues/15

**Comment by aholstenson** [comment]
@350d Thanks for adding some additional details! I don't think that plugin has been updated to be compatible with 0.15.x yet, some work is needed by the plugin author to support the new version. That might affect things for you.

**Comment by aholstenson** [comment]
@jghaanstra I've added a bit more debug info to the master-branch, if you have the possibility it would be great if you could test it out. The new debug-info should provide some more clues to how tokens are used and which ports is used to receive data.

I'm especially interested in the difference between the cases where you are able to connect and those where you are not.

**Comment by jghaanstra** [comment]
@aholstenson thank you for your efforts. Below are some logs again, one of a not working discovery attempt (DUMP 1) and one of a working connection using example.js and setting the token as option of miio.device() (DUMP 2).  I noticed that I was never able to connect using the local token storage. I have added tokens.json manually as `miio tokens update`  does not seem to work (DUMP 3) although I can see in the log it's using the correct token from tokens.json.

When setting the token as option of miio.device() manually in example.js I get a succes ratio of about 20%. I have added a log (DUMP 4) using the working example.js where it does not connect. The error suggests the token might be wrong but this is not the case as it works with this token sometimes.

Another strange thing, I have installed the latest Mi Home app and the token fields in miio2.db are all empty for my devices. It seems Xiaomi is gonna make it harder to control their devices through third party integration.

Anyway, here are the dumps:

**DUMP 1 - token registered in tokens.json and trying discovery through CLI or example.js (gives the same error)**
```
th:discovery:miio Searching every 100000 ms +0ms
  miio:network Grabbing reference to network +0ms
  miio:network Making network active, creating socket +0ms
  miio:network Network bound to port 57108 +24ms
  th:discovery:miio Searching for services +25ms
  miio:packet -> <Buffer 21 31 00 20 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 3a cc ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +158ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 3a cc ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +1ms
  thing:miio:51850703 <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  thing:miio:51850703 Loading token from storage, device hides token and no token set via options +0ms
  miio:tokens Loading token storage from C:\Users\jghaa\AppData\Local\miio\miio\tokens.json +0ms
  miio:tokens Loading tokens +1ms
  thing:miio:51850703 Using manual token: 4d707056323754564542566f34396274 +4ms
  thing:miio:51850703 -> (5) {"id":1,"method":"miIO.info","params":[]} +1ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a cc 80 91 d1 76 75 97 59 bd 56 b7 c6 c0 a3 62 20 99> +8ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 3a cd ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +603ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 3a cd ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +2ms
  thing:miio:51850703 <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +606ms
  thing:miio:51850703 -> (4) {"id":1,"method":"miIO.info","params":[]} +1s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a ce e8 f2 15 09 a5 7a 8a 81 0d b5 23 44 7a fb 08 85> +1s
  thing:miio:51850703 -> (3) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a d0 9a b0 01 b4 4c cd 9b ff fc 03 69 44 df 22 7d c1> +2s
  thing:miio:51850703 -> (2) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a d2 8d 7b a1 d2 3d 64 2c c5 5c 4d c8 45 f9 37 bc d0> +2s
  thing:miio:51850703 -> (1) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a d4 7d b0 ff 08 a7 d7 e5 67 2b d1 5c f7 65 dc ac 02> +2s
  thing:miio:51850703 -> (0) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a d6 b2 2e f9 06 47 8c cf ee 9b 84 9b b8 0e 55 05 00> +2s
  th:discovery:miio Service with id 51850703 now available +12s
  miio:network Grabbing reference to network +12s
  thing:miio:51850703 -> (5) {"id":2,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a d8 e5 2b 81 8a fc b4 a5 65 44 1c cb 6b 61 a2 8c c2> +2s
  thing:miio:51850703 -> (4) {"id":2,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a da 41 09 b7 36 d9 d3 0d b1 b8 e9 77 88 d0 06 01 9b> +2s
  thing:miio:51850703 -> (3) {"id":2,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a dc d0 07 d1 fa 19 54 42 34 c6 35 87 3c 01 ff 9c ff> +2s
  thing:miio:51850703 -> (2) {"id":2,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a de 88 fe bb d4 6f e3 09 26 c7 44 6a d6 8c 90 b4 e2> +2s
  thing:miio:51850703 -> (1) {"id":2,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a e0 d7 39 65 97 7a 0a 4e 59 46 48 4d 71 89 9d 1b 34> +2s
  thing:miio:51850703 -> (0) {"id":2,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3a e2 cb 8a 4f ff f3 94 41 c6 79 49 56 df 23 12 f1 85> +2s
  miio:network Releasing reference to network +12s
  th:discovery:mapped Could not map service { Error: Could not connect to device, token might be wrong
    at enrichPromise.promise.then.then.catch.err (E:\Localhost\bin\nodejs\node_modules\miio\lib\network.js:324:16)
    at <anonymous> code: 'connection-failure' } +0ms
```

**DUMP 2 - working connection with example.js and token as option in miio.device()**
```
  miio:network Grabbing reference to network +0ms
  miio:network Making network active, creating socket +2ms
  thing:miio:pending Using manual token: 4d707056323754564542566f34396274 +0ms
  miio:packet -> <Buffer 21 31 00 20 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  miio:network Network bound to port 61885 +10ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 3c f5 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +10ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 3c f5 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +3ms
  thing:miio:pending <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +16ms
  thing:miio:51850703 Identifier of device updated +0ms
  thing:miio:51850703 -> (5) {"id":1,"method":"miIO.info","params":[]} +1ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3c f5 b0 f0 1d e0 55 6e 2a a2 38 b5 54 16 5b cc 2b b6> +4ms
  miio:packet <- <Buffer 21 31 01 90 00 00 00 00 03 17 2d cf 5a 6a 3c f5 9f ac 32 c6 0e 17 3f b4 84 83 06 a3 b1 17 ea 68> +7ms
  miio:packet <- <Buffer 21 31 01 90 00 00 00 00 03 17 2d cf 5a 6a 3c f5 9f ac 32 c6 0e 17 3f b4 84 83 06 a3 b1 17 ea 68> +1ms
  thing:miio:51850703 <- Message: `{"partner_id":"","id":1,"code":0,"message":"ok","result":{"hw_ver":"Linux","fw_ver":"3.3.9_003094","ap":{"ssid":"xxx","bssid":"54:xx:xx:xx8:52:A8","rssi":-70},"netif":{"localIp":"192.168.0.133","mask":"255.255.255.0","gw":"192.168.0.100"},"model":"rockrobo.vacuum.v1","mac":"28:xx:xx:xx:AF:4B","token":"4d707056323754564542566f34396274","life":63972}}` +12ms
  miio:network Grabbing reference to network +23ms
  miio:network Releasing reference to network +3ms
  thing:miio:51850703 -> (5) {"id":2,"method":"get_status","params":[]} +5ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3c f5 87 0e 2e ae ee 42 e7 d4 39 9a ba fa fa 2b 5f 98> +9ms
  thing:miio:51850703 -> (5) {"id":3,"method":"get_consumable","params":[]} +2ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3c f5 23 2a 00 2b 68 d6 38 8b df 1e 18 0d e4 5d af f4> +2ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3c f5 93 dc bc 5c 85 a4 45 5a c7 f4 e1 f1 58 6e 97 1b> +7ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3c f5 93 dc bc 5c 85 a4 45 5a c7 f4 e1 f1 58 6e 97 1b> +2ms
  thing:miio:51850703 <- Message: `{"result":[{"msg_ver":6,"msg_seq":12,"state":8,"battery":100,"clean_time":1,"clean_area":0,"error_code":0,"map_present":0,"in_cleaning":0,"fan_power":90,"dnd_enabled":0}],"id":2}` +12ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3c f5 ae 4b 51 e4 37 2c 7f 22 b4 95 b0 56 29 50 d4 8e> +3ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3c f5 ae 4b 51 e4 37 2c 7f 22 b4 95 b0 56 29 50 d4 8e> +2ms
  thing:miio:51850703 <- Message: `{"result":[{"main_brush_work_time":469146,"side_brush_work_time":469146,"filter_work_time":469146,"sensor_dirty_time":47489}],"id":3}` +3ms
  thing:miio:51850703 Property state changed from undefined to charging +0ms
  thing:miio:51850703 Property batteryLevel changed from undefined to 100 +1ms
  thing:miio:51850703 Property cleanTime changed from undefined to 1 +1ms
  thing:miio:51850703 Property cleanArea changed from undefined to 0 +1ms
  thing:miio:51850703 Property fanSpeed changed from undefined to 90 +1ms
  thing:miio:51850703 Property in_cleaning changed from undefined to 0 +1ms
  thing:miio:51850703 Property mainBrushWorkTime changed from undefined to 469146 +0ms
  thing:miio:51850703 Property sideBrushWorkTime changed from undefined to 469146 +1ms
Connected to device
MiioDevice {
  model=rockrobo.vacuum.v1,
  types=miio:vacuum, miio, vaccuum,
  capabilities=adjustable-fan-speed, fan-speed, spot-cleaning, autonomous-cleaning, cleaning-state, error-state, charging-state, battery-level, state
}
  thing:miio:51850703 -> (5) {"id":4,"method":"get_status","params":[]} +30s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3d 12 19 f9 f6 5e 59 89 09 be 70 d0 71 54 32 e5 0f b0> +30s
  thing:miio:51850703 -> (5) {"id":5,"method":"get_consumable","params":[]} +4ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3d 12 b7 34 df 36 dc 15 c2 92 ca 6b f9 07 f3 52 84 df> +3ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3d 13 c4 c6 0e fe b2 c8 60 fc 49 8d fc 06 27 37 5f c3> +8ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3d 13 c4 c6 0e fe b2 c8 60 fc 49 8d fc 06 27 37 5f c3> +3ms
  thing:miio:51850703 <- Message: `{"result":[{"msg_ver":6,"msg_seq":30,"state":8,"battery":100,"clean_time":1,"clean_area":0,"error_code":0,"map_present":0,"in_cleaning":0,"fan_power":90,"dnd_enabled":0}],"id":4}` +13ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3d 13 c0 a5 ed db 5f 09 96 1a 5c ba d4 87 4a f7 29 5e> +1ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3d 13 c0 a5 ed db 5f 09 96 1a 5c ba d4 87 4a f7 29 5e> +1ms
  thing:miio:51850703 <- Message: `{"result":[{"main_brush_work_time":469146,"side_brush_work_time":469146,"filter_work_time":469146,"sensor_dirty_time":47489}],"id":5}` +2ms
  thing:miio:51850703 -> (5) {"id":6,"method":"get_status","params":[]} +30s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3d 30 46 69 9e 5e 44 c7 90 c4 b6 36 90 88 78 8c 59 fc> +30s
  thing:miio:51850703 -> (5) {"id":7,"method":"get_consumable","params":[]} +3ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3d 30 e0 8c 48 e8 fb 27 3c 2e f7 d6 fa 83 8f 9b af cd> +2ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3d 31 96 e8 56 2e 2a b1 38 bd bd cd 2a cd 41 e9 59 9e> +8ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3d 31 96 e8 56 2e 2a b1 38 bd bd cd 2a cd 41 e9 59 9e> +1ms
  thing:miio:51850703 <- Message: `{"result":[{"msg_ver":6,"msg_seq":31,"state":8,"battery":100,"clean_time":1,"clean_area":0,"error_code":0,"map_present":0,"in_cleaning":0,"fan_power":90,"dnd_enabled":0}],"id":6}` +11ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3d 31 35 52 0a 52 8c 45 0e e3 c6 b5 aa d2 e0 f5 d5 8d> +2ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3d 31 35 52 0a 52 8c 45 0e e3 c6 b5 aa d2 e0 f5 d5 8d> +1ms
  thing:miio:51850703 <- Message: `{"result":[{"main_brush_work_time":469146,"side_brush_work_time":469146,"filter_work_time":469146,"sensor_dirty_time":47489}],"id":7}` +3ms
  thing:miio:51850703 -> (5) {"id":8,"method":"get_status","params":[]} +30s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3d 4e 2e dd a2 15 18 5d 34 71 3a 54 5b 81 2a b4 42 1d> +30s
  thing:miio:51850703 -> (5) {"id":9,"method":"get_consumable","params":[]} +5ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 3d 4e 8a f8 2a e8 fa 3f 3a 7c c1 fc 9c 17 e5 85 f4 e9> +1ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3d 4f eb 11 b9 96 7b a9 67 92 86 7c b7 5a cd 2b 22 61> +10ms
  miio:packet <- <Buffer 21 31 00 e0 00 00 00 00 03 17 2d cf 5a 6a 3d 4f eb 11 b9 96 7b a9 67 92 86 7c b7 5a cd 2b 22 61> +2ms
  thing:miio:51850703 <- Message: `{"result":[{"msg_ver":6,"msg_seq":32,"state":8,"battery":100,"clean_time":1,"clean_area":0,"error_code":0,"map_present":0,"in_cleaning":0,"fan_power":90,"dnd_enabled":0}],"id":8}` +12ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3d 4f 82 09 a6 0f f1 31 31 07 38 29 42 db c5 54 1f ae> +1ms
  miio:packet <- <Buffer 21 31 00 b0 00 00 00 00 03 17 2d cf 5a 6a 3d 4f 82 09 a6 0f f1 31 31 07 38 29 42 db c5 54 1f ae> +1ms
  thing:miio:51850703 <- Message: `{"result":[{"main_brush_work_time":469146,"side_brush_work_time":469146,"filter_work_time":469146,"sensor_dirty_time":47489}],"id":9}` +3ms
```
**DUMP 3 - using miio tokens update**
```
λ miio tokens update 51850703 --token 4d707056323754564542566f34396274
 INFO  Updating token for 51850703
  th:discovery:miio Searching every 100000 ms +0ms
  miio:network Grabbing reference to network +0ms
  miio:network Making network active, creating socket +1ms
  miio:network Network bound to port 53289 +15ms
  th:discovery:miio Searching for services +18ms
  miio:packet -> <Buffer 21 31 00 20 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 40 6c ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +323ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 40 6c ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +3ms
  thing:miio:51850703 <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  thing:miio:51850703 Loading token from storage, device hides token and no token set via options +0ms
  miio:tokens Loading token storage from C:\Users\jghaa\AppData\Local\miio\miio\tokens.json +0ms
  miio:tokens Token storage does not exist +1ms
  th:discovery:miio Service with id 51850703 now available +339ms
  miio:network Grabbing reference to network +343ms
  miio:packet -> <Buffer 21 31 00 20 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +16ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 40 6c ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +4ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 40 6c ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +2ms
  thing:miio:51850703 <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +21ms
  thing:miio:51850703 Loading token from storage, device hides token and no token set via options +2ms
  miio:network Grabbing reference to network +9ms
  th:discovery:mapped Could not map service TypeError: Cannot read property 'id' of undefined
    at new <anonymous> (E:\Localhost\bin\nodejs\node_modules\miio\lib\device.js:38:34)
    at new module.exports (E:\Localhost\bin\nodejs\node_modules\miio\lib\placeholder.js:6:18)
    at network.findDeviceViaAddress.then.catch.e (E:\Localhost\bin\nodejs\node_modules\miio\lib\connectToDevice.js:35:12)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7) +0ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 40 6c ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +284ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 40 6c ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +2ms
  thing:miio:51850703 <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +284ms
 WARNING  Could not find device
```

**DUMP 4 - working example.js with token as option of miio.device() but no connection**
```
 miio:network Grabbing reference to network +0ms
  miio:network Making network active, creating socket +3ms
  thing:miio:pending Using manual token: 4d707056323754564542566f34396274 +0ms
  miio:packet -> <Buffer 21 31 00 20 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +0ms
  miio:network Network bound to port 64336 +11ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 41 63 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +10ms
  miio:packet <- <Buffer 21 31 00 20 00 00 00 00 03 17 2d cf 5a 6a 41 63 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +2ms
  thing:miio:pending <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +14ms
  thing:miio:51850703 Identifier of device updated +0ms
  thing:miio:51850703 -> (5) {"id":1,"method":"miIO.info","params":[]} +1ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 41 63 21 73 4a c3 23 58 88 a5 c3 33 d7 32 99 e8 44 fa> +4ms
  thing:miio:51850703 -> (4) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 41 65 a7 00 2b 21 bb ec a2 a8 4c f5 25 79 25 9d 8b 6e> +2s
  thing:miio:51850703 -> (3) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 41 67 dd cd c2 f5 ea 6f 9b f5 c4 a7 f9 17 33 6e de 38> +2s
  thing:miio:51850703 -> (2) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 41 69 29 ae 3f ca eb aa e8 2e 3e bd f2 cb 84 cc cb 1a> +2s
  thing:miio:51850703 -> (1) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 41 6b 00 0f e8 70 21 4e 43 8a 6c c1 e3 a8 2d 13 8c 17> +2s
  thing:miio:51850703 -> (0) {"id":1,"method":"miIO.info","params":[]} +2s
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6a 41 6d 1c ce 6c 00 11 7d cd cc 11 ad 67 26 a8 ce c7 16> +2s
  miio:network Releasing reference to network +12s
  miio:network Network no longer active, destroying socket +2ms
Error occurred: { Error: Could not connect to device, token might be wrong
    at enrichPromise.promise.then.then.catch.err (E:\Localhost\bin\nodejs\node_modules\miio\lib\network.js:324:16)
    at <anonymous> code: 'connection-failure' }
```

**Comment by aholstenson** [comment]
Thank you for taking the time to put that together! Sadly I can't quite figure out what's wrong, it just seems like the vacuum sometimes doesn't respond, almost like it's a sleep mode where it doesn't listen to the local network.

I've pushed 0.15.3 and 0.15.4 which contains some fixes to issues with token handling when they are being updated via the CLI. It shouldn't affect the overall connectivity though.

If anyone else is having similar issues or ideas about what is wrong, please do add details to this issue. Any information helps!

**Comment by jghaanstra** [comment]
I was afraid of that. Thanx for putting in the effort though. When I find some time I'll see if I have more success with the [python-miio](https://github.com/rytilahti/python-miio), a comparable library written in Python. If this shows similar behaviour I can pinpoint it on my network or configuration, if this works I can pinpoint it to the library and comparing them would be the next step.

One last question, on what Xiaomi server have you set up your Mi Robot?

**Comment by aholstenson** [comment]
Sounds good, comparing is always a good thing.

I think my vacuum is connected to the old US-server. I have also avoided updating the firmware, so I'm probably running a really old firmware by now.

**Comment by jghaanstra** [comment]
In retrospect I should have stayed on an older firmware too.

Could you explain a little what the difference is between the miio discovery and miio.device() in relation to the token? Discovery seems to always detect my device when I have not set the token in tokens.json, when I do it never finds my device. While miio.device() only connects works 5% of the time either using token in token.json or set it manually as option.

I'm trying to pinpoint where it fails. Looking at the log it doesnt seem to be a sleep issue to me but the robot just does not respond to `{"id":1,"method":"miIO.info","params":[]}` command most of the times and than destroys the socket. 

**WORKING miIO.info COMMAND**
```
  thing:miio:pending <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +74ms
  thing:miio:51850703 Identifier of device updated +0ms
  thing:miio:51850703 -> (5) {"id":1,"method":"miIO.info","params":[]} +1ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6b 9f bc 79 e2 d8 e4 32 33 c2 cd ab 69 fa 15 c1 c8 13 ff> +4ms
  miio:packet <- <Buffer 21 31 01 90 00 00 00 00 03 17 2d cf 5a 6b 9f bc 3e 32 eb 48 b8 cd f5 5b 43 f0 22 ec 69 df 95 25> +5ms
  miio:packet <- <Buffer 21 31 01 90 00 00 00 00 03 17 2d cf 5a 6b 9f bc 3e 32 eb 48 b8 cd f5 5b 43 f0 22 ec 69 df 95 25> +1ms
```

**NON WORKING miIO.info COMMAND**
```
 thing:miio:pending <- Handshake reply: <Buffer ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff> +26ms
  thing:miio:51850703 Identifier of device updated +0ms
  thing:miio:51850703 -> (5) {"id":1,"method":"miIO.info","params":[]} +1ms
  miio:packet -> <Buffer 21 31 00 50 00 00 00 00 03 17 2d cf 5a 6b 9e e9 9e 85 35 3e 0c f6 36 7f 77 a2 d8 cd 76 16 90 ef> +5ms
```

**Comment by aholstenson** [comment]
There is currently an issue with discovery treating connection failures as non-existent devices. I'm rewriting the discovery a bit right now, so in the next version it will return a "placeholder device" that can be used to update the token via code.

I've been looking at your dumps as well, did some manual testing to verify that the token has been set and that the checksums are correctly calculated with it. Everything checks out, so nothing new to report from my end, sorry.

I also read through rytilahti/python-miio/issues/92 where a few users seem to have the same connection issue. Most seems to have an issue with their token though, which isn't the case for you.

**Comment by jghaanstra** [comment]
Although the token is correct I'm not excluding it as a possible cause somehow. The current results of not connecting could for many users suggest the token is not correct (the actual error of your library even suggest so, and perhaps miio-python shows a similar token related error). I need to set up a python enviroment to test this in further detail, but I'm not to familiar with python so this will take me some time.

Something else I want to add to this issue for further troubleshooting. I have noticed that when I try to connect to the device it mostly works the first time after some cooling down periode. Trying to connect shortly after will not work. The suggests it's also not a sleep issue (but more an awake issue :-) ).

**Comment by jghaanstra** [comment]
Instead of setting up a new linux based enviroment I went with WinPython and have setup python-miio on my Windows machine. I can confirm that I do not have any issues with connecting to the Mi Robot using python-miio.

This pinpoints my connection issues to this specific miio library and the next step is comparing this library against the python-miio library. This is gonna be challinging for me as my javascript knowledge is limited and my python knowledge is pretty much non existent. But it's a step ahead ...

It would be great if more knowledgeable users with a similar issue could chip in though.

**Comment by aholstenson** [comment]
That's very interesting. Does running `python-miio` with debug output show anything interesting? Might provide a clue to the difference between the two libraries.

**Comment by jghaanstra** [comment]
Here is some debugging out:

```
INFO:miio.vacuum_cli:Debug mode active
DEBUG:miio.vacuum_cli:Read stored sequence ids: {'seq': 118, 'manual_seq': 0}
DEBUG:miio.vacuum_cli:Connecting to 192.168.0.133 with token 4d707056323754564542566f34396274
DEBUG:miio.protocol:Unable to decrypt, returning raw bytes: b''
DEBUG:miio.device:Got a response: Container:
    data = Container:
        data =  (total 0)
        value =  (total 0)
        offset1 = 32
        offset2 = 32
        length = 0
    header = Container:
        data = !1\x00 \x00\x00\x00\x00\x03\x17-\xcfZlh. (total 16)
        value = Container:
            length = 32
            unknown = 0
            device_id = 03172dcf (total 8)
            ts = 2018-01-27 11:53:18
        offset1 = 0
        offset2 = 16
        length = 16
    checksum = \xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff (total 16)
DEBUG:miio.device:Discovered b'03172dcf' with ts: 2018-01-27 11:53:18, token: b'ffffffffffffffffffffffffffffffff'
DEBUG:miio.device:192.168.0.133:54321 >>: {'id': 119, 'method': 'get_status'}
DEBUG:miio.device:192.168.0.133:54321 (ts: 2018-01-27 11:53:18, id: 119) << {'result': [{'msg_ver': 6, 'msg_seq': 13, 'state': 8, 'battery': 100, 'clean_time': 1, 'clean_area': 0, 'error_code': 0, 'map_present': 0, 'in_cleaning': 0, 'fan_power': 90, 'dnd_enabled': 0}], 'id': 119}
State: Charging
Battery: 100 %
Fanspeed: 90 %
Cleaning since: 0:00:01
Cleaned area: 0.0 m²
DEBUG:miio.vacuum_cli:Writing {'seq': 119, 'manual_seq': 0} to C:\Users\jghaa\AppData\Local\python-miio\python-miio\Cache/python-mirobo.seq
```

**Comment by aholstenson** [comment]
Thanks, I noticed the sequence numbers as well. I pushed a change to the repository does something similar on errors, could you see if that helps with the connection issues?

Regarding the state, I've added a few examples to the docs here: https://github.com/aholstenson/miio/blob/master/docs/devices/vacuum.md

I need to document the vacuum a lot more though, the state has been split into several different capabilities, such as one indicating if is charging, one if its cleaning, one if it has an error and so on. It's on my todo, going to try to get to soon.

**Comment by jghaanstra** [comment]
JACKPOT!!!

I'm able to connect now every time using the latest commit. I do notice the first connection is slightly faster then the following ones. I'm assuming your implementation of sequence numbers kicks in after it fails. Not too big a deal but it could be optimized to always use some sort of sequencing to avoid this issue. Thanx a million for figuring this out!

**Comment by aholstenson** [comment]
Awesome! 

The fixes have been pushed to NPM in version 0.15.5. Closing this issue and then we can see if we can do some optimization in a later version.
